DataFormats/Detectors/TRD/src/RawData.cxx

```diff
@@ -84,253 +103,265 @@ uint32_t getlinkerrorflag(const HalfCRUHeader& cruhead, const uint32_t link)
+      //pidword is here to make this code more readible and less error prone.
+      switch (pidindex) {
+        case 2: //Q2
+          if (!dynamicq) {
+            pid = (pidword >> 13) & 0x7f; // 7 bits at the top of all of pid (MSB)
+          } else {
+            pid = (pidword >> 12) & 0x3f; // 6 bits of Q2 and a shift
+            pid += pidoffset;
+          }
+          break;
+        case 1: //Q1
+          if (!dynamicq) {
+            pid = (pidword >> 6) & 0x7f; // 7 bits Q1 above the 6 bits of Q0
+          } else {
+            pid = (pidword >> 6) & 0x3f; // 6 bits of Q1 and a shift
+            pid += pidoffset;
+          }
+          break;
+        case 0: //Q0
+          if (!dynamicq) {
+            pid = pidword & 0x3f; // 6 least significant bits
+          } else {
+            pid = pidword & 0x3f; // 6 bits of Q0
+          }
+          break;
+        default:
+          LOG(warn) << " unknown pid index of : " << pidindex;
+          break;
+      }
+  } // end of case of various formats.
+  return pid;
 }
 
-void printTrackletMCMHeader(o2::trd::TrackletMCMHeader& mcmhead)
+//Tracklet MCM Header
+
+uint16_t constructTRDFeeID(int supermodule, int side, int endpoint)
 {
-  LOG(info) << " about to print mcm raw header";
-  LOGF(info, "MCMRawHeader: Raw:0x%08x 1:%d padrow: 0x%02x col: 0x%01x pid2 0x%02x pid1: 0x%02x pid0: 0x%02x 1:%d",
-       mcmhead.word, mcmhead.onea, mcmhead.padrow, mcmhead.col,
-       mcmhead.pid2, mcmhead.pid1, mcmhead.pid0, mcmhead.oneb);
-  LOG(info) << " printed mcm raw header";
+  TRDFeeID feeid;
+  feeid.word = 0;
+  feeid.supermodule = supermodule;
+  feeid.side = side;
+  feeid.endpoint = endpoint;
+  feeid.unused1 = 0;
+  feeid.unused2 = 0;
+  return feeid.word;
 }
 
-std::ostream& operator<<(std::ostream& stream, const TrackletMCMHeader& mcmhead)
+void constructTrackletMCMData(TrackletMCMData& trackletword, const int format, const uint slope, const uint pos, const uint q0, const uint q1, const uint q2)
 {
-  // make a pretty output of the mcm header.
-  stream << "TrackletMCMRawHeader: Raw:0x" << std::hex << mcmhead.word << " " << mcmhead.onea << "::"
-         << mcmhead.pid2 << ":" << mcmhead.pid1 << ":" << mcmhead.pid0 << "::"
-         << mcmhead.oneb << std::endl;
-  return stream;
+  trackletword.word = 0;
+  trackletword.slope = slope;
+  trackletword.pos = pos;
+  if (format & 0x1) {
+    //length of q1 and q2 are 6 with a 2 bit offset.
+    //What happens when q2 and q1 dont share the same offset (upper 2 bits ?) ?
+    LOG(error) << "This tracklet format has not been tested yet";
+    trackletword.pid = (q0 & 0x3f) & ((q1 & 0x1) << 6);
+    //TODO check the 2 bit offset is still valid ... ??
+  } else {
+    trackletword.pid = (q0 & 0x3f) & ((q1 & 0x1) << 6);
+  }
+  //q2 sits with upper 2 bits of q1 in the header pid word, hence the 0x1f so 5 bits are used here.
+  trackletword.checkbit = 1;
```

Comment:
Shouldn't this also be a trailing zero, or is the TDP here wrong as well?

Reply:
yes, but that function I need to fix in lieu of all these tracklet formats, else it will fail the fst.

Reply:
basically in fixing the acceptance of the tracklet formats. the format field in trap2raw needs to be properly done. I dont know that I am making this any clearer though :-)